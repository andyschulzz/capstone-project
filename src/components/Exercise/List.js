import React from 'react'
import Exercise from './Exercise'
import SearchBox from '../common/SearchBox'
import { Button } from '../common/Button'
import { useRouteMatch, Link } from 'react-router-dom'
import { useForm } from 'react-hook-form'
import * as S from './List.styles'
import PropTypes from 'prop-types'

List.propTypes = {
  exercises: PropTypes.array.isRequired,
  handleExerciseSelect: PropTypes.func.isRequired,
  handleSearch: PropTypes.func.isRequired,
  search: PropTypes.string,
}
const lettersAtoZ = [...Array(26)].map((_, i) =>
  String.fromCharCode('A'.charCodeAt(0) + i)
)

export default function List({
  exercises,
  handleExerciseSelect,
  handleSearch,
  search,
}) {
  const { url } = useRouteMatch()
  const { register, handleSubmit, reset } = useForm({ mode: 'onChange' })

  const renderedExercises = lettersAtoZ.reduce((acc, letter) => {
    const letterExercises = filterExercises(letter)
    if (letterExercises.length !== 0) {
      acc.push(<S.Span key={letter}>{letter}</S.Span>)
      letterExercises.forEach(lc => acc.push(lc))
    }
    return acc
  }, [])
  return (
    <>
      <S.ButtonWrapper>
        <Button primary="true" as={Link} to={`${url}/add`}>
          Add
        </Button>
        <SearchBox
          register={register}
          onSubmit={handleSubmit(onSubmit)}
          reset={reset}
          search={search}
        />
      </S.ButtonWrapper>
      <S.Wrapper>
        {exercises.length ? (
          renderedExercises
        ) : (
          <S.Container>No exercise found!</S.Container>
        )}
      </S.Wrapper>
    </>
  )

  function filterExercises(letter) {
    const filter = exercises
      .filter(exercise => String(exercise.name).startsWith(letter))
      .map((exercise, index) => {
        return (
          <Exercise
            key={exercise.id}
            {...exercise}
            index={index}
            handleExerciseSelect={handleExerciseSelect}
          />
        )
      })
    return filter
  }

  function onSubmit(userInput) {
    handleSearch(userInput)
  }
}
