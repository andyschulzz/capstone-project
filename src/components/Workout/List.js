import React from 'react'
import * as S from './List.styles'

export default function List({ workouts }) {
  const groupByTitle = workouts.reduce((acc, obj) => {
    const key = obj['title']
    if (!acc[key]) {
      acc[key] = []
    }
    acc[key].push(obj)
    return acc
  }, {})

  const renderWorkouts = Object.entries(groupByTitle).map(
    ([title, exercises]) => (
      <React.Fragment key={title}>
        <h3>{title}</h3>
        {exercises.map((exercise, index) => {
          return (
            <S.Wrapper key={index}>
              {exercise.name}
              <S.DetailsWrapper>
                {exercise.reps}
                <S.Span>&times;</S.Span>
                {exercise.sets}
                {exercise.weight && (
                  <S.Span isWeight>({exercise.weight} kg)</S.Span>
                )}
              </S.DetailsWrapper>
            </S.Wrapper>
          )
        })}
      </React.Fragment>
    )
  )
  return <>{renderWorkouts}</>
}
