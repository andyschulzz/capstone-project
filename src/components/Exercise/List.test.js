import React from 'react'
import { render, cleanup } from '@testing-library/react'
import List from './Exercise.js'
import { MemoryRouter as Router } from 'react-router-dom'

afterEach(cleanup)

const exercisesMock = [
  {
    id: 1,
    name: 'Test',
    type: 'Test',
    instructions: 'Test',
    title: 'Test',
  },
]

describe('<List />', () => {
  it('should render correctly', () => {
    const { asFragment } = render(
      <Router>
        <List exercises={exercisesMock} />
      </Router>
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
