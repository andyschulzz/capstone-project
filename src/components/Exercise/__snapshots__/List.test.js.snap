// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> should render correctly 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0 0 18px 10px;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 p:first-child {
  margin-top: 0;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  color: #333;
}

.c2 p:first-child {
  margin-top: 0;
}

.c3 {
  font-size: 1rem;
  margin: 10px 0 0 0;
}

.c4 {
  font-size: 0.8rem;
  margin: 0;
  font-weight: 300;
  color: #999;
}

.c1 {
  width: 46px;
  height: 46px;
  margin-right: 36px;
}

<a
    class="c0"
    href="//details/NaN"
  >
    <img
      alt=""
      class="c1"
      src="placeholder.png"
    />
    <div
      class="c2"
    >
      <p
        class="c3"
      />
      <p
        class="c4"
      />
    </div>
  </a>
</DocumentFragment>
`;
